File: Question8/Question8.py
# Participants may update the following function parameters
import itertools
def getMaximizedReturn(n, k, d, m, dailyReturns):
    og_ret = dailyReturns.copy()
    choices = k
    chosen = [False]*n 
    while choices>0:
        possible_loss = [0]*n
        profit_gained = [0]*n
        net=[0]*n
        max_net  = -float("inf")
        max_net_index = -1
        prefix_sum = []
        sum = 0
        for i in range(n):
            if dailyReturns[i]//m == og_ret[i] and dailyReturns[i]!=og_ret[i]:
                sum+= 0
            else:
                sum += dailyReturns[i]
            prefix_sum.append(sum)
        
        for i in range(n):
            if not chosen[i]: #available for selction 
               multiplier = max(0,(min([(k-1),d,i])))
               possible_loss[i] = (og_ret[i]*multiplier*m)# calc possible loss 
               profit_gained[i] = (prefix_sum[min(i+d,n-1)]-prefix_sum[i])*m
               net[i] = profit_gained[i]-possible_loss[i]-dailyReturns[i]
            if net[i]>max_net and not chosen[i]:
                max_net = net[i]
                max_net_index = i

        dailyReturns[max_net_index] = 0
        chosen[max_net_index] = True
        #print(min(max_net_index+1,n-1),min(max_net_index+d,n-1))
        for i in range(min(max_net_index+1,n-1),min(max_net_index+d,n-1)+1):
            if dailyReturns[i]//m != og_ret[i]:
                dailyReturns[i] *= m
        choices-=1
        
       # print("PS",prefix_sum)
       # print("OG",og_ret)
       # print("PL",possible_loss)
       # print("PG",profit_gained)
      #  print("NG",net)
       # print("choose",max_net_index)
       # print(dailyReturns)
       # print("next ")
   # print("final ",dailyReturns)
    sum = 0
    for i in dailyReturns:
        sum+=i
    return sum

def main():
    firstLine = input().split(" ")
    secondLine = input().split(" ")

    n = int(firstLine[0])
    k = int(firstLine[1])
    d = int(firstLine[2])
    m = int(firstLine[3])
    dailyReturns = list(map(int, secondLine))

    # Participants may update the following function parameters
    answer = getMaximizedReturn(n, k, d, m, dailyReturns)

    # Please do not remove the below line.
    print(answer)
    # Do not print anything after this line

if __name__ == '__main__':
    main()

File: Question8/Question8.cpp
//#include <bits/stdc++.h>
#include<iostream>
#include<vector>
#include<string>
using namespace std;


vector<int> splitStringToInt(const string& str, char delim) {
    vector<int> strings;
    size_t start;
    size_t end = 0;
    while ((start = str.find_first_not_of(delim, end)) != string::npos) {
        end = str.find(delim, start);
        strings.push_back(stoi(str.substr(start, end - start)));
    }
    return strings;
}


int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    string firstLine;
    getline(cin, firstLine);

    vector<int> firstLineVec = splitStringToInt(firstLine, ' ');
    int n = firstLineVec[0];
    int k = firstLineVec[1];
    int d = firstLineVec[2];
    int m = firstLineVec[3];

    string return_;
    getline(cin, return_);

    vector<int> returns = splitStringToInt(return_, ' ');
    int result = 0;
    vector<int> og_ret = returns;
    bool *chosen = new bool[n];
    for(int i = 0;i<n;i++){
        chosen[i] = false;
    }
    while(k>0){
        int pl[n];
        int pg[n];
        int net[n];
        int max_net = -__INT_MAX__;
        int mni = -1;
        int ps[n];
        int sum = 0;
        for(int i = 0;i<n;i++){
            if(og_ret[i]*m == returns[i] and og_ret[i] != returns[i]){
                sum+=0;
            }
            else{
                sum+=returns[i];
            }
            ps[i] = sum;
        }
        for(int i = 0;i<n;i++){
            if (!chosen[i]){
                int multiplier = min(i,(min(k-1,d)));
                pl[i] = og_ret[i]*multiplier*m;
                pg[i] = (ps[min(i+d,n-1)]-ps[i])*m;
                net[i] = pg[i]-(pl[i]+returns[i]);
                
            if(net[i]>max_net){
                max_net = net[i];
                mni = i;
                }
            }
        }
        returns[mni] = 0;
        chosen[mni] = true;
        for(int i = mni+1;i<=min(mni+d,n-1);i++){
            if(og_ret[i]*m!=returns[i] and !chosen[i]){
                returns[i]*= m;
           }
        }
    k--;
    }
    for(int i = 0 ;i <n;i++){
        result+=returns[i];
    } 
    // Do not remove below line
    cout << result;
    // Do not print anything after this line
    return 0;
}